lms/djangoapps/courseware/module_render.py

```diff
@@ -462,17 +474,16 @@ def inner_get_module(descriptor):
         )
 
     def publish(block, event_type, event):
-        """A function that allows XModules to publish events."""
-        if event_type == 'grade' and not is_masquerading_as_specific_student(user, course_id):
-            SCORE_PUBLISHED.send(
-                sender=None,
-                block=block,
-                user=user,
-                raw_earned=event['value'],
-                raw_possible=event['max_value'],
-                only_if_higher=event.get('only_if_higher'),
-                score_deleted=event.get('score_deleted'),
-            )
+        """
+        A function that allows XModules to publish events.
+        """
+        special_events = {
```

Comment:
Some questions to possibly lead to a more clear name:
What does `special_events` mean?  Are these just `events_with_handlers`?  Why do these particular events get handled here?

Reply:
1. They are special, just because they are handled in a particular way other than just passing them along to be logged.
2. Effectively, yes.
3. Because they are emitted from xblock javascript, and need to be handled by the backend, and this is the place where they come through.  I had suggested refactoring this into a system where you could register a handler with a particular event type, and it would get run here, which would basically allow the handlers to live with the grades/completion/progress apps.  The downside would be that it would spread the behavior around more, but the upside is that we could reduce the coupling to other apps.  Essentially, the registry could be a class with a signgle instance that has a register() method that attaches functions to event names, and then a send() method that sends events to the appropriate handler(s).